MGDAPI-4016 Added support to create MT RHOAM manifests and bundles

carlkyrillos opened the pull request:
# Issue link
JIRA: https://issues.redhat.com/browse/MGDAPI-4016

# What
With the onboarding of RHOAM to the DeveloperSandbox, we created a new installation type for RHOAM called `multitenant-managed-api`. This PR updates various scripts and support files to allow for the building of manifests and bundles for a new `OLM_TYPE` called `multitenant-managed-api-service`. More specifics and details can be found in the [JIRA issue](https://issues.redhat.com/browse/MGDAPI-4016). 

Note: The verification steps may look long but the instructions are just thorough. You'll basically just be building images and then triggering an upgrade.

# Verification steps
## Prerequisites
1. [Provision](https://qaprodauth.cloud.redhat.com/openshift/) a cluster
2. Create two [quay.io](https://quay.io/new/) repositories in your personal quay org using the following names (make sure to set them as **Public**)
   * multitenant-managed-api-service-bundle
   * multitenant-managed-api-service-index
## Create images for Multitenant RHOAM v1.22.0
1. Checkout this PR
2. Build and push bundle and index images: `ORG={personal-quay-org} OLM_TYPE=multitenant-managed-api-service UPGRADE=false BUNDLE_VERSIONS=1.22.0 make create/olm/bundle`
## Install Multitenant RHOAM v1.22.0
1. Prepare the cluster by running `INSTALLATION_TYPE=multitenant-managed-api make cluster/prepare/local` on the command line once the cluster has finished provisioning
2. Create the following CatalogSource either using `oc` or the GUI, (make sure to replace {ORG} with your quay org):
```yaml
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: rhmi-operators
  namespace: openshift-marketplace
spec:
  sourceType: grpc
  image: quay.io/{ORG}/multitenant-managed-api-service-index:1.22.0
```
3. Once the CatalogSource's Status is `READY`, install the RHOAM operator through the OperatorHub (OLM installation). Confirm that the Installed Namespace is `sandbox-rhoam-operator`
4. Once the operator is installed, create a RHMI CR by running `INSTALLATION_TYPE=multitenant-managed-api USE_CLUSTER_STORAGE=true make deploy/integreatly-rhmi-cr.yml` on the command line
5. The installation is complete when the `stage` field in the RHMI CR's Status block is `complete`. This can be checked using the GUI or by running `oc get rhmi rhoam -o json | jq .status.stage`
## Create images for Multitenant RHOAM v1.23.0
1. Create CSV for 1.23.0: `OLM_TYPE=multitenant-managed-api-service SEMVER=1.23.0 make release/prepare`
2. Build and push bundle and index images: `ORG={personal-quay-org} OLM_TYPE=multitenant-managed-api-service UPGRADE=true BUNDLE_VERSIONS=1.23.0,1.22.0 make create/olm/bundle`
## Upgrade Multitenant RHOAM from v1.22.0 to v1.23.0
1. Before continuing, confirm that the v1.22.0 installation has completed successfully
2. Once the installation is confirmed, trigger the upgrade by editing the CatalogSource you created earlier to point to `quay.io/{ORG}/multitenant-managed-api-service-index:1.23.0`
3. The upgrade should trigger automatically once you save the CatalogSource
4. Verify that the upgrade is successful by checking that the `version` field in the RHMI CR's Status block is `1.23.0`. This can be checked using the GUI or by running `oc get rhmi rhoam -o json | jq .status.version`
## Verify APIManagementTenant CRD works
1. Once the upgrade to v1.23.0 is complete, setup the testing IDP by running `DEDICATED_ADMIN_PASSWORD=Password1 PASSWORD=Password1 ./scripts/setup-sso-idp.sh` on the command line
2. Once the IDP is deployed on the cluster, login `test-user01` through the testing-idp so the User is created
3. Next create a new namespace/project called `test-user01-dev` using `oc` or the GUI
4. Then create an APIManagementTenant CR in the `test-user01-dev` namespace, example:
```yaml
apiVersion: integreatly.org/v1alpha1
kind: APIManagementTenant
metadata:
  name: example
  namespace: test-user01-dev
spec: {}
```
5. Verify that the APIManagementTenant CR reconciles (should take approximately 1 minute)

MStokluska commented on bundles/multitenant-managed-api-service/1.21.0/bundle.Dockerfile:
```diff
@@ -0,0 +1,21 @@
+LABEL operators.operatorframework.io.bundle.manifests.v1=manifests/
+LABEL operators.operatorframework.io.bundle.metadata.v1=metadata/
+LABEL operators.operatorframework.io.bundle.package.v1=multitenant-managed-api-service
+LABEL operators.operatorframework.io.bundle.channels.v1=stable
```
did we confirm that the channel name of the existing install is stable?

MStokluska replied:
I can confirm channel is stable

carlkyrillos replied:
Thank you for checking this! Marking it as resolved :+1:

MStokluska commented on bundles/multitenant-managed-api-service/1.21.0/bundle.Dockerfile:
```diff
@@ -0,0 +1,21 @@
+LABEL operators.operatorframework.io.bundle.mediatype.v1=registry+v1
+LABEL operators.operatorframework.io.bundle.manifests.v1=manifests/
+LABEL operators.operatorframework.io.bundle.metadata.v1=metadata/
+LABEL operators.operatorframework.io.bundle.package.v1=multitenant-managed-api-service
```
did we confirm that the package name of the existing install is multitenant-managed-api-service?

MStokluska replied:
It seems that existing package name is : ```"packageName":"managed-api-service","version":"1.20.0"}```

MStokluska replied:
Currently installed mt RHOAM is called managed-api-service.v1.20.0. The upgrade to this version will fail because of it

MStokluska commented on bundles/multitenant-managed-api-service/multitenant-managed-api-service.package.yaml:
```diff
@@ -0,0 +1,5 @@
+channels:
+  - currentCSV: multitenant-managed-api-service.v1.21.0
+    name: rhmi
+defaultChannel: rhmi
```
the channels seems to be different in metadata/annotations.yaml

carlkyrillos replied:
I was basing this off of single tenant RHOAM which has `stable` in [bundles/managed-api-service/1.22.0/metadata/annotations.yaml](https://github.com/integr8ly/integreatly-operator/blob/master/bundles/managed-api-service/1.22.0/metadata/annotations.yaml#L7-L8) but `rhmi` in [bundles/managed-api-service/managed-api-service.package.yaml](https://github.com/integr8ly/integreatly-operator/blob/master/bundles/managed-api-service/managed-api-service.package.yaml#L3-L4)

Should I change the channel to be `stable` in both or `rhmi` in both or keep it as is?

MStokluska commented on version/version.go:
```diff
@@ -12,9 +12,10 @@ const (
-	log               = l.NewLoggerWithContext(l.Fields{l.ComponentLogContext: "version"})
+	version                      = "2.9.0"
+	managedAPIVersion            = "1.22.0"
+	multitenantManagedAPIVersion = "1.21.0"
```
Hey Carl, is there any real need for this addition? 
I thought that an MT release would be the same as RHOAM release but with a slightly different setup on the bundle level.

carlkyrillos replied:
After discussing this on the main channel, we decided to keep this approach with separate versions for MT and standard RHOAM in order to simplify support for patch releases.

MStokluska commented on config/manifests-multitenant-rhoam/bases/multitenant-managed-api-service.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,143 @@
+metadata:
+  annotations:
+    operatorframework.io/suggested-namespace: sandbox-rhoam-operator
+    alm-examples: '[]'
```
Hey Carl, not sure where this warning is coming from 
```
WARN[0000] ClusterServiceVersion validation: [OperationFailed] provided API should have an example annotation 
```
But I guess it's because the newly generated CSV does not cover an example of tenants CRD

MStokluska commented on config/manifests-multitenant-rhoam/bases/multitenant-managed-api-service.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,143 @@
+    operators.operatorframework.io/project_layout: go.kubebuilder.io/v2
+    olm.properties: '[{"type": "olm.maxOpenShiftVersion", "value": "4.10"}]'
+    support: RHOAM
+  name: multitenant-managed-api-service.v1.21.0
```
Currently installed mt RHOAM is called managed-api-service.v1.20.0. The upgrade to this version will fail because of it

MStokluska commented:
hey @carlkyrillos so I've verified upgrade from the current mt installation 1.20 > 1.21 works after a tweak to the package name value and some comments left around the code verification.

MStokluska commented on config/manifests-multitenant-rhoam/bases/multitenant-managed-api-service.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,143 @@
+      type: SingleNamespace
+    - supported: false
+      type: MultiNamespace
+    - supported: false
```
it's ns scoped. Sorry.

briangallagher commented on bundles/multitenant-managed-api-service/1.21.0/manifests/multitenant-managed-api-service.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,747 @@
+    categories: Integration & Delivery
+    certified: "false"
+    containerImage: quay.io/integreatly/multitenant-managed-api-service:v1.21.0
+    olm.properties: '[{"type": "olm.maxOpenShiftVersion", "value": "4.10"}]'
```
@carlkyrillos Can we change this to 4.10.15 and also look at how we might keep this up to date. It should be the same as what we've tested on nightly pipelines. Maybe it's part of the release process and manual.

carlkyrillos replied:
Yes will do :+1:

briangallagher replied:
@carlkyrillos 4.10.15 actually won't work. OLM will complain. So, 4.10 is fine .

briangallagher commented on Makefile:
```diff
@@ -32,6 +32,7 @@ SANDBOX_NAMESPACE ?= sandbox-rhoam-operator
 # These tags are modified by the prepare-release script.
 RHMI_TAG ?= 2.9.0
 RHOAM_TAG ?= 1.22.0
+MULTITENANT_RHOAM_TAG ?= 1.21.0
```
@carlkyrillos Similar to Michals previous comment, I'm not sure there is a need for a different version tag? 
I think maybe the issue is that we should be building 1.21.0 based on the 1.21.0 code base and not the 1.22.0 code base. So, it's more of a process issue.

carlkyrillos replied:
After discussing this on the main channel, we decided to keep this approach with separate versions for MT and standard RHOAM in order to simplify support for patch releases.

carlkyrillos commented:
@MStokluska @briangallagher I pushed a new commit that addresses most of the feedback and rebased off of master.

@MStokluska When you get a chance could you take another look at the changes? One thing to note, after talking with Brian we decided to keep the package name as `multitenant-managed-api-service` despite current MT rhoam using `managed-api-service`. We determined this was acceptable because the only existing installations of MT rhoam are on our DevSanbox environments and we can just manually tweak the names when performing the next upgrade. After we do this once, it won't be an issue going forward.

carlkyrillos commented:
/test gosec

MStokluska commented on scripts/prepare-release.sh:
```diff
@@ -142,29 +155,39 @@ set_descriptions() {
       yq e -i '.spec.maintainers[0].name="rhoam"' bundles/$OLM_TYPE/${VERSION}/manifests/$OLM_TYPE.clusterserviceversion.yaml
       ;;
+    "multitenant-managed-api-service")
+      echo "Updating permissions"
```
would it make more sense to say here "updating maintainers" rather than permissions?

carlkyrillos replied:
Yeah I think that makes more sense, updated in the latest push.

carlkyrillos commented on bundles/multitenant-managed-api-service/multitenant-managed-api-service.package.yaml:
```diff
@@ -0,0 +1,5 @@
+  - currentCSV: multitenant-managed-api-service.v1.22.0
+    name: rhmi
+defaultChannel: rhmi
+packageName: multitenant-managed-api-service
```
I think it makes sense to just fix it now, I updated the values in the latest push.

MStokluska commented on bundles/multitenant-managed-api-service/bundle.Dockerfile:
```diff
@@ -0,0 +1,17 @@
+LABEL operators.operatorframework.io.bundle.mediatype.v1=registry+v1
+LABEL operators.operatorframework.io.bundle.manifests.v1=manifests/
+LABEL operators.operatorframework.io.bundle.metadata.v1=metadata/
+LABEL operators.operatorframework.io.bundle.package.v1=multitenant-managed-api-service
```
This will need to change to `managed-api-service` otherwise the package is still `multitenant-managed-api-service` and upgrade fails.

carlkyrillos replied:
Sorry not sure how I missed this but it's been fixed in the latest push.

MStokluska commented on bundles/multitenant-managed-api-service/1.22.0/bundle.Dockerfile:
```diff
@@ -0,0 +1,21 @@
+FROM scratch
```
no need for this file to be in.

carlkyrillos replied:
Is this file also unnecessary for standard versions of RHOAM? Right now, `scripts/prepare-release.sh` copies `bundle.Dockfile` into the `bundles/${OLM_TYPE}/${VERSION}/` directory but if neither version of RHOAM needs it, I'll get rid of the `cp` command from the prepare-release.sh script

MStokluska commented on config/manifests-multitenant-rhoam/bases/multitenant-managed-api-service.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,143 @@
+    operators.operatorframework.io/project_layout: go.kubebuilder.io/v2
+    olm.properties: '[{"type": "olm.maxOpenShiftVersion", "value": "4.10"}]'
+    support: RHOAM
+  name: multitenant-managed-api-service.v1.22.0
```
```suggestion
  name: managed-api-service.v1.22.0
```

MStokluska commented on scripts/prepare-release.sh:
```diff
@@ -17,9 +17,14 @@ case $OLM_TYPE in
     ;;
+  "multitenant-managed-api-service")
+    PREVIOUS_VERSION=$(grep $OLM_TYPE bundles/$OLM_TYPE/$OLM_TYPE.package.yaml | awk -F v '{print $3}') || echo "No previous version"
+    PACKAGE_NAME=multitenant-managed-api-service
```
do we want to bump this too to managed-api-service?

MStokluska replied:
Is there a need for the csv name to be multitenant managed api?
The only difference that I can see between RHOAM and MT RHOAM is a couple of rbacs + additional CRD, is there any harm in having same CSV name?

carlkyrillos replied:
No I don't think there's any harm in having the same CSV name. I changed it to reflect this in the latest push.

carlkyrillos commented:
/test rhoam-e2e
/test multitenant-rhoam-e2e

MStokluska commented:
@carlkyrillos I would actually add 1.22 too. The reason why is because you/or some ci/cd system, will have to build the bundles and we also know for sure that 1.23 will have to replace 1.22. Without 1.22 as a starting point, we will either have to remove replaces from 1.23 (which is going to break the upgrade) or point to custom-built 1.22 that will sit somewhere in someone's private repo. (alternatively, add olm.SkipRange to 1.23)

carlkyrillos commented:
/test rhoam-e2e
/test multitenant-rhoam-e2e

MStokluska commented:
Reviewed upgrade from 1.22 (current MT) > 1.23 > 1.24, works as expected. Good job!
/lgtm

MStokluska commented:
/lgtm

briangallagher commented on version/version.go:
```diff
@@ -12,9 +12,10 @@ const (
-	version           = "2.9.0"
-	managedAPIVersion = "1.24.0"
-	log               = l.NewLoggerWithContext(l.Fields{l.ComponentLogContext: "version"})
+	version                      = "2.9.0"
```
@carlkyrillos @MStokluska I think we can remove the changes to this file. In general MT RHOAM will use the same operator image as standard rhoam. In the rare occasions where we need to install and emergency fix to MT RHOAM master will already by set to n+1. Also, I think this process can be manual (custom image) given for sandbox

carlkyrillos replied:
Addressed in latest commit.

briangallagher commented on scripts/prepare-release.sh:
```diff
@@ -45,133 +45,77 @@ else
-          yq e -i '(.spec.install.spec.deployments[0].spec.template.spec.containers[0].env.[] | select(.name=="INSTALLATION_TYPE") | .value) = "managed-api" ' bundles/$OLM_TYPE/${VERSION}/manifests/$OLM_TYPE.clusterserviceversion.yaml
+        "multitenant-managed-api-service")
+          "${SED_INLINE[@]}" -E "s/MULTITENANT_RHOAM_TAG\s+\?=\s+$PREVIOUS_VERSION/MULTITENANT_RHOAM_TAG \?= $VERSION/g" Makefile
+          "${SED_INLINE[@]}" -E "s/multitenantManagedAPIVersion\s+=\s+\"$PREVIOUS_VERSION\"/multitenantManagedAPIVersion = \"$VERSION\"/g" version/version.go
```
@carlkyrillos I don't think we need this, see comment below

carlkyrillos replied:
Addressed in latest commit

briangallagher commented:
I added 2 comments re. version.go. Everything else looks good

MStokluska commented:
hey @carlkyrillos checked the latest version, it seems that when I ran the `ORG=mstoklus OLM_TYPE=multitenant-managed-api-service UPGRADE=true BUNDLE_VERSIONS=1.23.0,1.22.0 make create/olm/bundle` the index image for 1.23 had only 1.23 bundle in it. Any chance you could look into it?

carlkyrillos commented:
> hey @carlkyrillos checked the latest version, it seems that when I ran the `ORG=mstoklus OLM_TYPE=multitenant-managed-api-service UPGRADE=true BUNDLE_VERSIONS=1.23.0,1.22.0 make create/olm/bundle` the index image for 1.23 had only 1.23 bundle in it. Any chance you could look into it?

@MStokluska I'm unable to replicate the behavior you're describing; when I run `ORG=ckyrillo OLM_TYPE=multitenant-managed-api-service UPGRADE=true BUNDLE_VERSIONS=1.23.0,1.22.0 make create/olm/bundle` both bundles are in the index. I ran that command against a clean version of this PR with no other changes.

I checked the bundles by first running the index as a container: `docker run --rm -it -p 50051:50051 quay.io/ckyrillo/multitenant-managed-api-service-index:1.23.0` 

Then in a separate terminal I inspected the contents of the index using grpcurl: `grpcurl -plaintext localhost:50051 api.Registry/ListBundles | jq '{packageName: .packageName, name: .csvName, version: .version, bundlePath: .bundlePath, replaces: .replaces, skipRange: .skipRange, channelName: .channelName, dependencies: .dependencies}'` and here is the output:
```
{
  "packageName": "managed-api-service",
  "name": "managed-api-service.v1.22.0",
  "version": "1.22.0",
  "bundlePath": "quay.io/ckyrillo/multitenant-managed-api-service-bundle:1.22.0",
  "replaces": null,
  "skipRange": null,
  "channelName": "stable",
  "dependencies": null
}
{
  "packageName": "managed-api-service",
  "name": "managed-api-service.v1.23.0",
  "version": "1.23.0",
  "bundlePath": "quay.io/ckyrillo/multitenant-managed-api-service-bundle:1.23.0",
  "replaces": "managed-api-service.v1.22.0",
  "skipRange": null,
  "channelName": "stable",
  "dependencies": null
}
```

I need to rebase and remove the version changes per @briangallagher comments but then could you try running it again?

carlkyrillos commented:
@MStokluska @briangallagher I rebased and removed the separate versioning for multitenant RHOAM. I also went through the verification steps again and confirmed that everything is working as expected. 

@MStokluska Could you try running `ORG=mstoklus OLM_TYPE=multitenant-managed-api-service UPGRADE=true BUNDLE_VERSIONS=1.23.0,1.22.0 make create/olm/bundle` again and see if the latest changes fix things on your end?

carlkyrillos commented:
/test rhoam-e2e
/test multitenant-rhoam-e2e

MStokluska commented:
Thanks @carlkyrillos 
/lgtm

briangallagher commented:
/approve

carlkyrillos commented:
/test rhoam-e2e
/test multitenant-rhoam-e2e

briangallagher commented:
/lgtm
/approve
